### PR TITLE
Code with refs

### PR DESCRIPTION
**Pull request by danking**

### CODE REVIEW

File: src/main/scala/is/hail/asm4s/CodeWithRefs.scala [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package is.hail.asm4s
+
+import scala.language.implicitConversions
+
+case class CodeWithRefs[T](action: FunctionBuilder[_] => T) {
```
Comment (cseed) [inline src/main/scala/is/hail/asm4s/CodeWithRefs.scala]: Rename.  CodeWithRefs sounds like it should extend/contain a Code.  It's really FunctionBuilderReader.  That's not great, either.  It's the thing that actually generates the byte code.  CodeGen?

File: src/main/scala/is/hail/asm4s/CodeWithRefs.scala [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  def map[U](f: T => U): CodeWithRefs[U] =
+    CodeWithRefs(fb => f(action(fb)))
+  def flatMap[U](f: T => CodeWithRefs[U]): CodeWithRefs[U] =
+    CodeWithRefs(fb => f(action(fb)).action(fb))
+  def withFilter(p: T => Boolean): CodeWithRefs[T] =
+    CodeWithRefs(fb => { val t = action(fb); if (p(t)) t else throw new RuntimeException() })
+  def filter(p: T => Boolean): CodeWithRefs[T] =
+    withFilter(p)
+  def run[R >: Null](fb: Function0Builder[R])(implicit ev: T <:< Code[R]): () => R =
```
Comment (cseed) [inline src/main/scala/is/hail/asm4s/CodeWithRefs.scala]: We need to generalize the result/run interface so we don't have to specialize it for each type.  That can go in a separate PR if you like.

File: src/test/scala/is/hail/asm4s/CodeWithRefsSuite.scala [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+object CodeWithRefsSuite {
+  def toISeq(a: Array[Int]): IndexedSeq[Int] = a
+}
+
+class CodeWithRefsSuite extends TestNGSuite {
+
+  @Test def sum() {
+    val m = for {
+      a <- newVar[Array[Int]](Code.newArray[Int](3))
+      _ <- a.update(0, 1)
```
Comment (cseed) [inline src/test/scala/is/hail/asm4s/CodeWithRefsSuite.scala]: We can't use := here?  Oh, there is no ArrayRef.  We should add that.  (Separate PR OK.)

**Comment by danking** [comment]
@cseed comments addressed in the other PR that has arguments
